File: ChatRoom.Client/Network/ServerMessageBag.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace ChatRoom.Client.Network
{
    public class ServerMessageBag
    {
        private int recode;
        private string message = string.Empty;
        private string type = string.Empty;
        private JObject data;

        public int Recode { get => recode; set => recode = value; }
        public string Message { get => message; set => message = value; }
        public string Type { get => type; set => type = value; }
        public JObject Data { get => data; set => data = value; }

        public ServerMessageBag(string jsonString)
        {
            JObject obj = JsonConvert.DeserializeObject<JObject>(jsonString) ?? new JObject();

            recode = obj.Value<int>("recode");
            message = obj.Value<string>("msg") ?? string.Empty;
            type = obj.Value<string>("type") ?? string.Empty;
            data = obj.Value<JObject>("data") ?? new JObject();
        }
    }
}

File: ChatRoom.Server/ServerMessageBag.cpp
#include "ServerMessageBag.h"
using namespace ServerMessage;

ServerMessageBag::ServerMessageBag(const string& type)
{
	_recode = 0;
	_message = "";
	_type = type;
}

ServerMessageBag ServerMessageBag::AddRecode(int recode, const string& message)
{
	_recode = recode;
	_message = message;

	return *this;
}

string ServerMessageBag::ToJsonString() const
{
	json j;
	j["type"] = _type;
	j["recode"] = _recode;
	j["msg"] = _message;
	j["data"] = _data;
	return j.dump();
}

template<typename T>
ServerMessageBag ServerMessageBag::AddData(const string& key, const T& value)
{
	_data[key] = value;
	return *this;
}